### PR TITLE
improve error handling for GDrive downloads

### PR DESCRIPTION
**Pull request by pmeier**
We have plenty of reports that the download of datasets hosted on GDrive does not work as expected:

- #1920
- #2262
- #2317
- #2756
- #3708
- #4615
- #5615
- #5716
- #5800

Although most of them are closed, they see another comment from time to time since a user stumbled upon "the same issue"

The errors non-descriptive in most cases. The problem is that we don't check the MD5 sum after the download and naively write the response from GDrive to disk. In contrast, on `download_url` we perform such a check

https://github.com/pytorch/vision/blob/aa21197462591a89f527ded3938ffe62f7d8cb8f/torchvision/datasets/utils.py#L150-L152

The most common failure case is GDrive returning an unknown API response as HTML. This PR adds a MD5 check after the download with an additional HTML check to make the error message more descriptive.

### CODE REVIEW

File: torchvision/datasets/utils.py [hunk 1]
```diff
@@ -257,6 +257,27 @@ def download_file_from_google_drive(file_id: str, root: str, filename: Optional[
...
+        msg = f"The MD5 checksum of the download file {fpath} does not match the one on record."
+        remediation_msg = (
+            "Please delete the file and try again. "
+            "If the issue persists, please report this to torchvision at https://github.com/pytorch/vision/issues."
+        )
+        # GDrive API responses should be smaller than 10kB
+        if os.stat(fpath).st_size < 10 * 1024:
+            with open(fpath) as fh:
+                content = fh.read()
+                # Regular expression to detect HTML. Copied from https://stackoverflow.com/a/70585604
```
Comment (NicolasHug) [inline torchvision/datasets/utils.py:270]: thank you for writing this comment

**Review by NicolasHug — COMMENTED** [review]
Thanks @pmeier . 

> The most common failure case is GDrive returning an unknown API response as HTML

Shouldn't we check the HTML code of the download / reply from GDrive instead of downloading then?

**Comment by pmeier** [comment]
> Shouldn't we check the HTML code of the download / reply from GDrive instead of downloading then?

Unfortunately, GDrive mostly just returns `200` and thus there is very little we can do without actually inspecting the response. Maybe we can improve this by always analyzing the first chunk of data rather than only if we detect an MD5 mismatch. Thoughts?

**Comment by NicolasHug** [comment]
> analyzing the first chunk of data rather than only if we detect an MD5 mismatch. Thoughts

If you mean checking against the HTML regex right after the files are downloaded, then yeah I feel like this might be where the check needs to be

**Review by NicolasHug — APPROVED** [review]
Thanks Philip

**Pull request merged**